lib/deep_cover/autoload_tracker.rb

```diff
@@ -0,0 +1,212 @@
+        end
+      end
+    end
+
+    # We need to remove the interceptor hooks, otherwise, the problem if manually requiring
+    # something that is autoloaded will cause issues.
+    def remove_interceptors(&do_autoload_block)
+      @autoloads_by_basename.each do |basename, entries|
+        entries.each do |entry|
+          mod = entry.mod_if_available
+          next unless mod
+          # Module's constants are shared with Object. But if you set autoloads directly on Module, they
+          # appear on multiple classes. So just skip, Object will take care of those.
+          next if mod == Module
+          yield mod, entry.name, entry.target_path
+        end
+      end
+
+      @autoloaded_paths = {}
+      @interceptor_files_by_path = {}
+    end
+
+    class << self
+      attr_accessor :warned_for_frozen_module
+    end
+    self.warned_for_frozen_module = false
+
+    # Using frozen modules/classes is almost unheard of, but a warning makes things easier if someone does it
+    def self.warn_frozen_module(mod)
+      return if warned_for_frozen_module
+      self.warned_for_frozen_module ||= true
+      warn "There is an autoload on a frozen module/class: #{mod}, DeepCover cannot handle those, failure is probable. " \
+           "This warning won't be displayed again (even for different module/class)"
+    end
+
+    protected
+
+    def setup_interceptor_for(mod, name, path)
+      interceptor_path = autoload_interceptor_for(path)
+      entry = AutoloadEntry.new(WeakRef.new(mod), name, path, interceptor_path)
+
+      basename = basename_without_extension(path)
+
+      @autoloads_by_basename[basename] ||= []
+      @autoloads_by_basename[basename] << entry
+      interceptor_path
+    end
+
+    def entries_for_target(requested_path, absolute_path_found)
+      basename = basename_without_extension(requested_path)
+      autoloads = @autoloads_by_basename[basename] || []
+
+      if absolute_path_found
+        autoloads.select { |entry| entry_is_target?(entry, requested_path, absolute_path_found) }
+      elsif requested_path == File.absolute_path(requested_path)
+        []
+      elsif requested_path.start_with?('./', '../')
+        []
+      else
+        # We didn't find a path that goes through the $LOAD_PATH
+        # It's possible that RubyGems will actually add the $LOAD_PATH and require an actual file
+        # So we must make a best-guest for possible matches
+        requested_path_to_compare = without_extension(requested_path)
+        autoloads.select { |entry| requested_path_to_compare == without_extension(entry.target_path) }
+      end
+    end
+
+    def entry_is_target?(entry, requested_path, absolute_path_found)
+      return true if entry.target_path == requested_path
+      target_path_rb = with_rb_extension(entry.target_path)
+      return true if target_path_rb == requested_path
+      entry_absolute_path = DeepCover.custom_requirer.resolve_path(entry.target_path)
```

Comment:
I'd add the following comment:
> Even though this is not efficient, it's best to resolve entries' target_path each time instead of storing the result, in case subsequent changes to $LOAD_PATH gives different results

(Or else I'm missing something?)

Reply:
Done